src/meta/meta_ckpt.c

```diff
@@ -918,9 +918,14 @@ __wt_meta_sysinfo_set(WT_SESSION_IMPL *session)
         WT_ERR(__wt_metadata_update(session, WT_SYSTEM_CKPT_URI, buf->data));
     }
 
-    /* We also need to record the oldest timestamp in the metadata so we can set it on startup. */
-    __wt_timestamp_to_hex_string(
-      S2C(session)->txn_global.checkpoint_oldest_timestamp, hex_timestamp);
+    /*
+     * We also need to record the oldest timestamp in the metadata so we can set it on startup. We
+     * should set the checkpoint's oldest timestamp as the minimum of the current oldest timestamp
+     * and the checkpoint timestamp.
+     */
+    __wt_timestamp_to_hex_string(WT_MIN(S2C(session)->txn_global.oldest_timestamp,
+                                   S2C(session)->txn_global.meta_ckpt_timestamp),
```

Comment:
This function now uses `S2C` in a number of places - it's better to make a local `conn` variable.